File: frontend/src/services/applicationService.ts
import { api } from './api';

export interface CreateApplicationRequest {
  job_id: string;
  cover_letter: string;
  custom_resume?: string;
}

export interface UpdateApplicationRequest {
  status?: string;
  cover_letter?: string;
  custom_resume?: string;
  employer_feedback?: string;
  candidate_feedback?: string;
  internal_notes?: string;
}

export interface ApplicationStatusUpdateRequest {
  status: string;
  feedback?: string;
  interview_scheduled?: string;
  interview_notes?: string;
}

export interface ApplicationWithdrawalRequest {
  reason?: string;
}

export interface ApplicationResponse {
  id: string;
  job_id: string;
  user_id: string;
  status: string;
  cover_letter: string;
  custom_resume?: string;
  match_score: number;
  applied_at: string;
  reviewed_at?: string;
  response_at?: string;
  employer_feedback?: string;
  candidate_feedback?: string;
  internal_notes?: string;
  interview_scheduled?: string;
  interview_notes?: string;
  created_at: string;
  updated_at: string;
  job?: JobSummaryResponse;
  applicant?: UserSummaryResponse;
}

export interface JobSummaryResponse {
  id: string;
  title: string;
  company: string;
  location: string;
  salary: string;
  duration: string;
  category: string;
  is_remote: boolean;
  status: string;
}

export interface UserSummaryResponse {
  id: string;
  name: string;
  email: string;
  user_type: string;
  experience_level: string;
  location: string;
  bio: string;
  skills: string[];
  level: number;
  xp: number;
}

export interface ApplicationListResponse {
  applications: ApplicationResponse[];
  total: number;
  page: number;
  limit: number;
  total_pages: number;
}

export interface ApiResponse<T = unknown> {
  success: boolean;
  data?: T;
  message?: string;
  errors?: string[];
}

class ApplicationService {
  // Submit a new application
  async submitApplication(request: CreateApplicationRequest): Promise<ApiResponse<ApplicationResponse>> {
    try {
      const response = await api.post('/applications', request);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to submit application',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get application by ID
  async getApplication(id: string): Promise<ApiResponse<ApplicationResponse>> {
    try {
      const response = await api.get(`/applications/${id}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get application',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Update application
  async updateApplication(id: string, request: UpdateApplicationRequest): Promise<ApiResponse<ApplicationResponse>> {
    try {
      const response = await api.put(`/applications/${id}`, request);
      return {
        success: true,
        data: response.data
      };
    } catch (error: unknown) {
      return {
        success: false,
        error: error.response?.data?.error || 'Failed to update application'
      };
    }
  }

  // Delete application
  async deleteApplication(id: string): Promise<ApiResponse> {
    try {
      await api.delete(`/applications/${id}`);
      return {
        success: true,
        message: 'Application deleted successfully'
      };
    } catch (error: unknown) {
      return {
        success: false,
        error: error.response?.data?.error || 'Failed to delete application'
      };
    }
  }

  // Get user applications
  async getUserApplications(page: number = 1, limit: number = 20, status?: string): Promise<ApiResponse<ApplicationListResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });
      
      if (status) {
        params.append('status', status);
      }

      const response = await api.get(`/applications/user?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get user applications',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Get job applications (for employers)
  async getJobApplications(jobId: string, page: number = 1, limit: number = 20, status?: string): Promise<ApiResponse<ApplicationListResponse>> {
    try {
      const params = new URLSearchParams({
        page: page.toString(),
        limit: limit.toString()
      });
      
      if (status) {
        params.append('status', status);
      }

      const response = await api.get(`/jobs/${jobId}/applications?${params.toString()}`);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to get job applications',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Update application status (for employers)
  async updateApplicationStatus(id: string, request: ApplicationStatusUpdateRequest): Promise<ApiResponse<ApplicationResponse>> {
    try {
      const response = await api.put(`/applications/${id}/status`, request);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to update application status',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Withdraw application
  async withdrawApplication(id: string, request: ApplicationWithdrawalRequest = {}): Promise<ApiResponse> {
    try {
      const response = await api.put(`/applications/${id}/withdraw`, request);
      return response.data;
    } catch (error: any) {
      return {
        success: false,
        message: error.response?.data?.message || 'Failed to withdraw application',
        errors: error.response?.data?.errors || [error.message]
      };
    }
  }

  // Check if user has already applied to a job
  async checkExistingApplication(jobId: string): Promise<ApiResponse<ApplicationResponse | null>> {
    try {
      const response = await api.get(`/applications/job/${jobId}/check`);
      return {
        success: true,
        data: response.data
      };
    } catch (error: unknown) {
      if (error.response?.status === 404) {
        return {
          success: true,
          data: null // No existing application
        };
      }
      return {
        success: false,
        error: error.response?.data?.error || 'Failed to check existing application'
      };
    }
  }

  // Get application statistics for a user
  async getUserApplicationStats(): Promise<ApiResponse<{
    total: number;
    pending: number;
    reviewed: number;
    accepted: number;
    rejected: number;
    withdrawn: number;
  }>> {
    try {
      const response = await api.get('/applications/user/stats');
      return {
        success: true,
        data: response.data
      };
    } catch (error: unknown) {
      return {
        success: false,
        error: error.response?.data?.error || 'Failed to get application statistics'
      };
    }
  }

  // Get application statistics for a job (employer)
  async getJobApplicationStats(jobId: string): Promise<ApiResponse<{
    total: number;
    pending: number;
    reviewed: number;
    accepted: number;
    rejected: number;
    withdrawn: number;
  }>> {
    try {
      const response = await api.get(`/applications/job/${jobId}/stats`);
      return {
        success: true,
        data: response.data
      };
    } catch (error: unknown) {
      return {
        success: false,
        error: error.response?.data?.error || 'Failed to get job application statistics'
      };
    }
  }
}

export const applicationService = new ApplicationService();
export default applicationService;

File: backend/internal/services/application_service.go
package services

import (
	"context"
	"fmt"
	"time"

	"microbridge/backend/internal/dto"
	"microbridge/backend/internal/models"
	"microbridge/backend/internal/repository"
	apperrors "microbridge/backend/internal/shared/errors"

	"github.com/google/uuid"
)

type ApplicationService interface {
	SubmitApplication(ctx context.Context, userID string, req dto.CreateApplicationRequest) (*dto.ApplicationResponse, error)
	GetApplication(ctx context.Context, id string) (*dto.ApplicationResponse, error)
	UpdateApplication(ctx context.Context, applicationID string, req dto.UpdateApplicationRequest) (*dto.ApplicationResponse, error)
	WithdrawApplication(ctx context.Context, applicationID, userID string, req dto.ApplicationWithdrawalRequest) error
	GetUserApplications(ctx context.Context, userID string, page, limit int) (*dto.PaginatedApplicationResponse, error)
	GetJobApplications(ctx context.Context, jobID string, employerID string, page, limit int) (*dto.PaginatedApplicationResponse, error)
	GetApplicationDetails(ctx context.Context, applicationID string, requesterID string) (*dto.ApplicationResponse, error)
	UpdateApplicationStatus(ctx context.Context, applicationID, employerID string, req dto.ApplicationStatusUpdateRequest) (*dto.ApplicationResponse, error)
}

type applicationService struct {
	applicationRepo repository.ApplicationRepository
	jobRepo         repository.JobRepository
	userRepo        repository.UserRepository
}

func NewApplicationService(
	applicationRepo repository.ApplicationRepository,
	jobRepo repository.JobRepository,
	userRepo repository.UserRepository,
) ApplicationService {
	return &applicationService{
		applicationRepo: applicationRepo,
		jobRepo:         jobRepo,
		userRepo:        userRepo,
	}
}

func (s *applicationService) SubmitApplication(ctx context.Context, userID string, req dto.CreateApplicationRequest) (*dto.ApplicationResponse, error) {
	// Validate the request
	if err := s.validateCreateApplicationRequest(req); err != nil {
		return nil, err
	}

	// Check if job exists and is active
	job, err := s.jobRepo.GetByID(ctx, req.JobID)
	if err != nil {
		return nil, err
	}

	if job.Status != "posted" {
		return nil, apperrors.NewAppError(400, "Job is not accepting applications", nil)
	}

	// Check if user already applied for this job
	existingApplication, err := s.applicationRepo.GetByUserAndJob(ctx, userID, req.JobID)
	if err != nil && !apperrors.IsNotFoundError(err) {
		return nil, err
	}
	if existingApplication != nil {
		return nil, apperrors.NewAppError(409, "You have already applied for this job", nil)
	}

	// Get user profile for matching
	user, err := s.userRepo.GetByID(ctx, userID)
	if err != nil {
		return nil, err
	}

	// Create application
	application := &models.Application{
		ID:           uuid.New().String(),
		UserID:       userID,
		JobID:        req.JobID,
		Status:       "submitted",
		CoverLetter:  req.CoverLetter,
		CustomResume: req.CustomResume,
		MatchScore:   s.calculateMatchScore(user, job), // Simplified for now
		AppliedAt:    time.Now(),
		CreatedAt:    time.Now(),
		UpdatedAt:    time.Now(),
	}

	if err := s.applicationRepo.Create(ctx, application); err != nil {
		return nil, err
	}

	// Update job application count
	job.Applications++
	if err := s.jobRepo.Update(ctx, job); err != nil {
		// Log error but don't fail the application creation
	}

	return s.applicationToResponse(application, job, user), nil
}

func (s *applicationService) GetApplication(ctx context.Context, id string) (*dto.ApplicationResponse, error) {
	application, err := s.applicationRepo.GetByID(ctx, id)
	if err != nil {
		return nil, err
	}

	// Get job and user details
	job, err := s.jobRepo.GetByID(ctx, application.JobID)
	if err != nil {
		return nil, err
	}

	user, err := s.userRepo.GetByID(ctx, application.UserID)
	if err != nil {
		return nil, err
	}

	return s.applicationToResponse(application, job, user), nil
}

func (s *applicationService) UpdateApplication(ctx context.Context, applicationID string, req dto.UpdateApplicationRequest) (*dto.ApplicationResponse, error) {
	application, err := s.applicationRepo.GetByID(ctx, applicationID)
	if err != nil {
		return nil, err
	}

	// Update fields if provided
	if req.Status != "" {
		application.Status = req.Status
	}
	if req.CoverLetter != "" {
		application.CoverLetter = req.CoverLetter
	}
	if req.CustomResume != "" {
		application.CustomResume = req.CustomResume
	}
	if req.EmployerFeedback != "" {
		application.EmployerFeedback = req.EmployerFeedback
	}
	if req.CandidateFeedback != "" {
		application.CandidateFeedback = req.CandidateFeedback
	}
	if req.InternalNotes != "" {
		application.InternalNotes = req.InternalNotes
	}

	application.UpdatedAt = time.Now()

	if err := s.applicationRepo.Update(ctx, application); err != nil {
		return nil, err
	}

	// Get job and user details
	job, err := s.jobRepo.GetByID(ctx, application.JobID)
	if err != nil {
		return nil, err
	}

	user, err := s.userRepo.GetByID(ctx, application.UserID)
	if err != nil {
		return nil, err
	}

	return s.applicationToResponse(application, job, user), nil
}

func (s *applicationService) WithdrawApplication(ctx context.Context, applicationID, userID string, req dto.ApplicationWithdrawalRequest) error {
	application, err := s.applicationRepo.GetByID(ctx, applicationID)
	if err != nil {
		return err
	}

	// Check if the user owns this application
	if application.UserID != userID {
		return apperrors.NewAppError(403, "You don't have permission to withdraw this application", nil)
	}

	// Check if application can be withdrawn
	if application.Status == "accepted" || application.Status == "rejected" {
		return apperrors.NewAppError(400, "Cannot withdraw application that has been finalized", nil)
	}

	// Update application status
	application.Status = "withdrawn"
	application.CandidateFeedback = req.Reason
	application.UpdatedAt = time.Now()

	return s.applicationRepo.Update(ctx, application)
}

func (s *applicationService) GetUserApplications(ctx context.Context, userID string, page, limit int) (*dto.PaginatedApplicationResponse, error) {
	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 20
	}

	offset := (page - 1) * limit

	applications, total, err := s.applicationRepo.GetByUserID(ctx, userID, limit, offset)
	if err != nil {
		return nil, err
	}

	applicationResponses := make([]*dto.ApplicationResponse, len(applications))
	for i, app := range applications {
		// Get job details
		job, err := s.jobRepo.GetByID(ctx, app.JobID)
		if err != nil {
			// If job is deleted, show minimal info
			job = &models.Job{
				ID:       app.JobID,
				Title:    "Job No Longer Available",
				Company:  "Unknown",
				Location: "Unknown",
				Status:   "archived",
			}
		}

		user, err := s.userRepo.GetByID(ctx, app.UserID)
		if err != nil {
			continue // Skip if user not found
		}

		applicationResponses[i] = s.applicationToResponse(app, job, user)
	}

	return &dto.PaginatedApplicationResponse{
		Applications: applicationResponses,
		Pagination: dto.PaginationResponse{
			Page:    page,
			Limit:   limit,
			Total:   total,
			HasMore: int64(page*limit) < total,
		},
	}, nil
}

func (s *applicationService) GetJobApplications(ctx context.Context, jobID string, employerID string, page, limit int) (*dto.PaginatedApplicationResponse, error) {
	// Verify that the employer owns this job
	job, err := s.jobRepo.GetByID(ctx, jobID)
	if err != nil {
		return nil, err
	}

	if job.EmployerID != employerID {
		return nil, apperrors.NewAppError(403, "You don't have permission to view these applications", nil)
	}

	if page <= 0 {
		page = 1
	}
	if limit <= 0 || limit > 100 {
		limit = 20
	}

	offset := (page - 1) * limit

	applications, total, err := s.applicationRepo.GetByJobID(ctx, jobID, limit, offset)
	if err != nil {
		return nil, err
	}

	applicationResponses := make([]*dto.ApplicationResponse, len(applications))
	for i, app := range applications {
		user, err := s.userRepo.GetByID(ctx, app.UserID)
		if err != nil {
			continue // Skip if user not found
		}

		applicationResponses[i] = s.applicationToResponse(app, job, user)
	}

	return &dto.PaginatedApplicationResponse{
		Applications: applicationResponses,
		Pagination: dto.PaginationResponse{
			Page:    page,
			Limit:   limit,
			Total:   total,
			HasMore: int64(page*limit) < total,
		},
	}, nil
}

func (s *applicationService) GetApplicationDetails(ctx context.Context, applicationID string, requesterID string) (*dto.ApplicationResponse, error) {
	application, err := s.applicationRepo.GetByID(ctx, applicationID)
	if err != nil {
		return nil, err
	}

	// Get job and verify permission
	job, err := s.jobRepo.GetByID(ctx, application.JobID)
	if err != nil {
		return nil, err
	}

	// Check if requester has permission (either applicant or job employer)
	if application.UserID != requesterID && job.EmployerID != requesterID {
		return nil, apperrors.NewAppError(403, "You don't have permission to view this application", nil)
	}

	user, err := s.userRepo.GetByID(ctx, application.UserID)
	if err != nil {
		return nil, err
	}

	return s.applicationToResponse(application, job, user), nil
}

func (s *applicationService) UpdateApplicationStatus(ctx context.Context, applicationID, employerID string, req dto.ApplicationStatusUpdateRequest) (*dto.ApplicationResponse, error) {
	application, err := s.applicationRepo.GetByID(ctx, applicationID)
	if err != nil {
		return nil, err
	}

	// Verify that the employer owns this job
	job, err := s.jobRepo.GetByID(ctx, application.JobID)
	if err != nil {
		return nil, err
	}

	if job.EmployerID != employerID {
		return nil, apperrors.NewAppError(403, "You don't have permission to update this application", nil)
	}

	// Update application
	application.Status = req.Status
	application.EmployerFeedback = req.Feedback
	application.InterviewScheduled = req.InterviewScheduled
	application.InterviewNotes = req.InterviewNotes
	application.UpdatedAt = time.Now()

	// Set response timestamp for final decisions
	if req.Status == "accepted" || req.Status == "rejected" {
		now := time.Now()
		application.ResponseAt = &now
	}

	if err := s.applicationRepo.Update(ctx, application); err != nil {
		return nil, err
	}

	user, err := s.userRepo.GetByID(ctx, application.UserID)
	if err != nil {
		return nil, err
	}

	return s.applicationToResponse(application, job, user), nil
}

// Helper methods

func (s *applicationService) validateCreateApplicationRequest(req dto.CreateApplicationRequest) error {
	if req.JobID == "" {
		return apperrors.NewAppError(400, "Job ID is required", nil)
	}
	if req.CoverLetter == "" {
		return apperrors.NewAppError(400, "Cover letter is required", nil)
	}
	if len(req.CoverLetter) < 10 {
		return apperrors.NewAppError(400, "Cover letter must be at least 10 characters", nil)
	}

	return nil
}

func (s *applicationService) calculateMatchScore(user *models.User, job *models.Job) float64 {
	// Simplified matching score calculation
	// In a real implementation, this would use a sophisticated matching algorithm
	score := 0.5 // Base score

	// Skill matching (simplified)
	userSkillMap := make(map[string]bool)
	for _, skill := range user.Skills {
		userSkillMap[skill.Name] = true
	}

	matchedSkills := 0
	totalSkills := len(job.Skills)
	for _, reqSkill := range job.Skills {
		if userSkillMap[reqSkill.Name] {
			matchedSkills++
		}
	}

	if totalSkills > 0 {
		skillScore := float64(matchedSkills) / float64(totalSkills)
		score = (score + skillScore) / 2
	}

	// Experience level matching
	if user.ExperienceLevel == job.ExperienceLevel {
		score += 0.1
	}

	// Location matching
	if user.Location == job.Location || job.IsRemote {
		score += 0.1
	}

	// Ensure score is between 0 and 1
	if score > 1.0 {
		score = 1.0
	}
	if score < 0.0 {
		score = 0.0
	}

	return score
}

func (s *applicationService) applicationToResponse(application *models.Application, job *models.Job, user *models.User) *dto.ApplicationResponse {
	response := &dto.ApplicationResponse{
		ID:                 application.ID,
		JobID:              application.JobID,
		UserID:             application.UserID,
		Status:             application.Status,
		CoverLetter:        application.CoverLetter,
		CustomResume:       application.CustomResume,
		MatchScore:         application.MatchScore,
		AppliedAt:          application.AppliedAt,
		ReviewedAt:         application.ReviewedAt,
		ResponseAt:         application.ResponseAt,
		EmployerFeedback:   application.EmployerFeedback,
		CandidateFeedback:  application.CandidateFeedback,
		InternalNotes:      application.InternalNotes,
		InterviewScheduled: application.InterviewScheduled,
		InterviewNotes:     application.InterviewNotes,
		CreatedAt:          application.CreatedAt,
		UpdatedAt:          application.UpdatedAt,
	}

	// Add job summary
	if job != nil {
		salaryStr := ""
		if job.Salary.Min > 0 || job.Salary.Max > 0 {
			if job.Salary.Min == job.Salary.Max {
				salaryStr = fmt.Sprintf("%s %d %s", job.Salary.Currency, job.Salary.Min, job.Salary.Period)
			} else {
				salaryStr = fmt.Sprintf("%s %d-%d %s", job.Salary.Currency, job.Salary.Min, job.Salary.Max, job.Salary.Period)
			}
		}

		response.Job = &dto.JobSummaryResponse{
			ID:       job.ID,
			Title:    job.Title,
			Company:  job.Company,
			Location: job.Location,
			Salary:   salaryStr,
			Duration: fmt.Sprintf("%d weeks", job.Duration),
			Category: job.Category,
			IsRemote: job.IsRemote,
			Status:   job.Status,
		}
	}

	// Add user summary
	if user != nil {
		skillNames := make([]string, len(user.Skills))
		for i, skill := range user.Skills {
			skillNames[i] = skill.Name
		}

		response.Applicant = &dto.UserSummaryResponse{
			ID:              user.ID,
			Name:            user.Name,
			Email:           user.Email,
			UserType:        user.UserType,
			ExperienceLevel: user.ExperienceLevel,
			Location:        user.Location,
			Bio:             user.Bio,
			Skills:          skillNames,
			Level:           user.Level,
			XP:              user.XP,
		}
	}

	return response
}